text/0000-layout-packed-aligned.md

```diff
@@ -0,0 +1,176 @@
+
+This RFC enables the following struct definitions:
+
+```rs
+#[repr(C, packed(2), align(4))]
+struct Foo { // Alignment = 4, Size = 8
+    a: u8,   // Offset = 0
+    b: u32,  // Offset = 2
+}
+```
+
+This is commonly needed when Rust is being used to interop with existing C and C++ code bases, which may contain
+unaligned types. For example in `clang` it is possible to create the following type definition, and there is
+currently no easy way to create a matching Rust type:
+
+```cpp
+struct  __attribute__((packed, aligned(4))) MyStruct {
+    uint8_t a;
+    uint32_t b;
+};
+```
+
+Currently `#[repr(packed(_))]` structs cannot transitively contain #[repr(align(_))] structs due to differing behavior between msvc and gcc/clang.
+However, in most cases, the user would expect `#[repr(C)]` to produce a struct layout matching the same type as defined by the current target. 
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## `#[repr(C)]`
+When `align` and `packed` attributes exist on the same type, or when `packed` structs transitively contains `align` types,
+the resulting layout matches the target toolchain ABI.
+
+For example, given:
+```c
+#[repr(C, align(4))]
+struct Foo(u8);
+#[repr(C, packed(1))]
+struct Bar(Foo);
+```
+`align_of::<Bar>()` would be 4 for `*-pc-windows-msvc` and 1 for everything else.
+
+
+## `#[repr(system)]`
+When `align` and `packed` attributes exist on the same type, or when `packed` structs transitively contains `align` types,
+the resulting layout matches the target OS ABI.
+
+For example, given:
+```c
+#[repr(system, align(4))]
+struct Foo(u8);
+#[repr(system, packed(1))]
+struct Bar(Foo);
+```
+`align_of::<Bar>()` would be 4 for `*-pc-windows-msvc` and `*-pc-windows-gnu`. It would be 1 for everything else.
+
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+In the following paragraphs, "Decreasing M to N" means:
+```
+if M > N {
+    M = n
+}
+```
+
+"Increasing M to N" means:
+```
+if M < N {
+    M = N
+}
+```
+
+
+`#[repr(align(N))]` increases the base alignment of a type to be N.
+
+`#[repr(packed(M))]` decreases the alignment of the struct fields to be M. Because the base alignment of the type
+is defined as the maximum of the alignment for any fields, this also has the indirect result of decreasing the base
+alignment of the type to be M.
+
+When the align and packed modifiers are applied on the same type as `#[repr(align(N), packed(M))]`,
+the alignment of the struct fields are decreased to be M. Then, the base alignment of the type is
+increased to be N.
+
+When a `#[repr(packed(M))]` struct transitively contains a field with `#[repr(align(N))]` type,
+- The field is first `pad_to_align`. Then, the field is added to the struct with alignment decreased to M. The packing requirement overrides the alignment requirement. (GCC, `#[repr(Rust)]`, `#[repr(C)]` on gnu targets, `#[repr(system)]` on non-windows targets)
```

Comment:
However, Clang and GCC has the following layout:

```c
struct  __attribute__((aligned(4)))
MyStructInner { // Alignment = 4, Size = 4
    uint8_t a; // Offset = 0
};

struct  __attribute__((packed))
MyStruct { // Alignment = 1, Size = 8
    MyStructInner a; // Offset = 0
    uint32_t b; // Offset = 4
};
```

It is indeed curious that `b` would have offset = 4 in `MyStruct`, even though it's not technically necessary. This behavior was covered by the language I used in the RFC.

Reply:
No, there's no curiosity here? `MyStructInner` has a size of 4, so the next field to follow it must be offset by at least 4 bytes from the field of type `MyStructInner` to avoid overlapping it. It's not allowed to put fields within the trailing padding of other fields in C, C++, nor Rust. (Unless using `[[no_unique_address]]` in C++.)

Reply:
I think I might understand where the disconnect is coming from. `MyStruct` *does not* have the field of type `uint8_t` pushed into its layout; it has a field of type `MyStructInner` pushed into its layout, and that `MyStructInner` has the full size of 4 bytes.

No additional alignment needs to be applied to push the field, because the `MyStructInner` type is already sufficiently sized and aligned for the `MyStructInner` value.

Reply:
I don't even know what "the field is first `pad_to_align`" even means. First of all I can't quite parse it from a grammatical PoV, and secondly the output of a layout algorithm is the offset of each field of the type (that is, the immediate fields listed in the type declaration, not the recursive ones). So it's completely unclear what this business with recursive fields here is supposed to mean. It sounds like it is talking about the offset of a field of an inner type in the outer type (like, the offset of a field of `MyStructInner` in `MyStruct`), but that makes no sense.

Reply:
Ok I see where the disconnect came from as well. I was under the impression that `MyStructInner` would have a size of 1 byte and an alignment of 4 bytes. If `MyStructInner` already have a size of 4 bytes, `pad_to_align` is indeed not necessary, although it wouldn't be incorrect.